### PR TITLE
Prevent rapid repeated closing of all connections

### LINKED ISSUE #65954: apiserver cannot recover after restarting apiserver in large scale cluster(5k nodes, 15w pods)
On a large cluster - 5k nodes and 15w pods(more than 500MB for pod spec). 5 master nodes, there are one apiserver and etcd on each master nodes.
When restart all the apiservers, the memory of etcd members increases rapidly to about 100G. And the leader keep to be elected frequently. And in the etcd log, it shows some request takes very long time.
```
2018-07-02 17:17:43.986312 W | etcdserver: apply entries took too long [52.361859051s for 1 entries]
2018-07-02 17:17:43.986341 W | etcdserver: avoid queries with large range/delete range!
```
After investigating, kubelet and other daemonsets on each nodes use listwatch all pods with nodeName fieldSelector(the first LIST request is send with resourceVersion=0 and nodename fieldSelector). When the pod watchcache of apiserver is not ready, apiserver will forward these LIST request to etcd. Because the pods count is very large, it will take about almost 60s for pod watchcache to become ready. Also when apiserver restart, all the pod watch from kubelet and daemonsets will resend LIST request in parallel. And it will be forward to etcd, it's a huge pressure to etcd. As to the LIST request was forward to etcd, the behavior is trying to fix the issue (https://github.com/kubernetes/kubernetes/issues/49684), (https://github.com/kubernetes/kubernetes/pull/49688) .

I tried to change this behavior in my test, let it return err when watch cache is not ready. Then before cache is ready, all these listwatch will receive 500 status code. After cache is ready, all the LIST request will send in parallel to apiserver again, although the status of etcd cluster is normal, the stress is put to apiserver. From the pprof, it's busy on filter pods on specified node from all the 15w pods.
```
File: kube-apiserver
Type: cpu
Time: Jul 5, 2018 at 5:26am (-07)
Duration: 30.14s, Total samples = 3.73mins (742.04%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 164.39s, 73.51% of 223.62s total
Dropped 935 nodes (cum <= 1.12s)
Showing top 10 nodes out of 96
      flat  flat%   sum%        cum   cum%
    40.78s 18.24% 18.24%    181.28s 81.07%  k8s.io/kubernetes/pkg/registry/core/pod.GetAttrs
    25.50s 11.40% 29.64%    119.42s 53.40%  runtime.mapassign
    23.53s 10.52% 40.16%     23.53s 10.52%  runtime.procyield
    17.91s  8.01% 48.17%     17.91s  8.01%  runtime.memclrNoHeapPointers
    12.11s  5.42% 53.59%     12.11s  5.42%  runtime.futex
    11.35s  5.08% 58.66%     11.35s  5.08%  runtime.heapBitsForObject
       11s  4.92% 63.58%     27.96s 12.50%  runtime.scanobject
     8.84s  3.95% 67.53%      8.84s  3.95%  runtime.memmove
     7.12s  3.18% 70.72%      7.16s  3.20%  runtime.greyobject
     6.25s  2.79% 73.51%      6.25s  2.79%  runtime.memeqbody
(pprof) tree pod.GetAttrs
Showing nodes accounting for 176.74s, 79.04% of 223.62s total
Dropped 69 nodes (cum <= 1.12s)
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context
----------------------------------------------------------+-------------
                                           181.28s   100% |   k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.(*SelectionPredicate).Matches
    40.78s 18.24% 18.24%    181.28s 81.07%                | k8s.io/kubernetes/pkg/registry/core/pod.GetAttrs
                                           119.07s 65.68% |   runtime.mapassign
                                            19.35s 10.67% |   runtime.makemap
                                             2.07s  1.14% |   runtime.writebarrierptr
----------------------------------------------------------+-------------
                                           119.07s   100% |   k8s.io/kubernetes/pkg/registry/core/pod.GetAttrs
    25.47s 11.39% 29.63%    119.07s 53.25%                | runtime.mapassign
                                            78.31s 65.77% |   runtime.newarray
                                            12.77s 10.72% |   runtime.typedmemmove
                                             1.85s  1.55% |   runtime.aeshashbody
                                             0.24s   0.2% |   runtime.writebarrierptr
```
In apiserver log, filter takes a lot of time, which occurs in parallel LIST(with fieldselector).
```
I0706 07:17:07.904668       7 trace.go:76] Trace[1077488077]: "cacher *api.Pod: List" (started: 2018-07-06 07:16:43.341190891 +0000 UTC) (total time: 24.563439891s):
Trace[1077488077]: [14.682341166s] [14.68234064s] watchCache locked acquired
Trace[1077488077]: [15.232549756s] [550.208282ms] Listed 148431 items from cache
Trace[1077488077]: [24.563432917s] [9.330883161s] Filtered 33 items
I0706 07:17:07.906590       7 trace.go:76] Trace[1612848086]: "cacher *api.Pod: List" (started: 2018-07-06 07:16:28.629546457 +0000 UTC) (total time: 39.277009968s):
Trace[1612848086]: [2.257410319s] [2.257409725s] watchCache locked acquired
Trace[1612848086]: [39.277001772s] [37.003565308s] Filtered 33 items
```
The impact is that:
the liveness probe of apiserver was configured
```
    livenessProbe:
      failureThreshold: 3
      httpGet:
        httpHeaders:
        path: /healthz
        port: 6443
        scheme: HTTPS
      initialDelaySeconds: 60
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 15
```
the liveness probe will fail and apiserver was keeping to killed and restart.
When I change the `initialDelaySeconds` to 300s, all the listwatch can succeed finally, althrough the latency is more than 30s. But the watchcache of some apiservers cannot refresh in time. kubelet connect to these apiservers keep to patch node status conflict. Lots of nodes became NotReady. In apisever log, the messages as below show up.
```
W0706 09:57:27.988402       7 watcher.go:236] Fast watcher, slow processing. Number of buffered events: 100.Probably caused by slow dispatching events to watchers
```
```
W0706 09:50:26.808609       7 watcher.go:343] Fast watcher, slow processing. Number of buffered events: 100.Probably caused by slow decoding, user not receiving fast, or other processing logic
```

There will be a large scale of this kind of LIST, if it do filter for each one, especially when the total resource count is large and the query is in parallel, it will waste a lot of time. 
Is there a index mechanism to get a filtered list directly? the obj was stored in a index structure? If so, it will save lots of filter time, mitigate the pressure to apiserver.

**Environment**:
- 40 core for each apiserver. the cpu usage can reach to 4000%.
- Kubernetes version (use `kubectl version`):  release 1.9
- max in-flight: 4000.
- max mutating in-flight: 200 (default)

@smarterclayton @wojtek-t

### PR DESCRIPTION
**Pull request by JohnRusk**
Such closing puts more load on the API server, leading to more heartbeat
failures, and then more closing.  This change adds a circuit breaker
for that cycle, so we can't close all the connections too often.

Fixes #65954

#### What type of PR is this?

/kind bug

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by JohnRusk** [comment]
BTW, the logic of why this is a safe change is as follows:

(a) The behaviour it modifies consists of resetting connections, and was specifically introduced to solve #48638, which is about "stuck" TCP connections.  If those are left unaddressed for too long (e.g. 15 minutes) it causes problems as documented in that issue.
(b) The modification made by this PR prevents connections from being reset any more often than once every 5 minutes.  
(c) From reading #48638, I undertand that the original need will still be more than adequately addressed by resets every 5 minutes.  @derekwaynecarr , as the author of that issue, would you agree?

**Comment by JohnRusk** [comment]
cc @jackfrancis As discussed, this PR is for the issue we've come to know as the "kubelet list storm"

**Comment by aojea** [comment]
is this happening with http2? it has a 30 seconds idle timeout and uses ping frames, so it will detect stale connections

**Comment by aojea** [comment]
> is this happening with http2? it has a 30 seconds idle timeout and uses ping frames, so it will detect stale connections

ok, I've found now the other comment https://github.com/kubernetes/kubernetes/issues/65954#issuecomment-1021929791 ,

I have to read it ,

**Comment by JohnRusk** [comment]
>is this happening with http2? it has a 30 seconds idle timeout and uses ping frames, so it will detect stale connections

This is with http2. But the closing of the connections is not due to failed pings.  It is specifically due to  this sequence of events:

-- copied from a comment I left on #65954 --

1. Immediately after restarting, API server(s) are busy. This is to be expected, because there's a lot of activity from kubelets when the API server restarts. But...
2. Due to being very busy, the API server sometimes takes more than 10 seconds to reply to some requests. But...
3. Kubelet designates certain operations as "heartbeat" operations, and gives them a 10 second timeout. ([Here ](https://github.com/kubernetes/kubernetes/blob/dbb5e3fe861c233d2f633b174497916e29ee68d8/pkg/kubelet/kubelet_node_status.go#L458) and [here](https://github.com/kubernetes/kubernetes/blob/dbb5e3fe861c233d2f633b174497916e29ee68d8/staging/src/k8s.io/component-helpers/apimachinery/lease/controller.go#L194)).  
4. If two of those heartbeat operations fail in a row, kubelet calls a function named `onRepeatedHeartbeatFailure` which _closes all TCP connections_.  
4. When the connections close, kubelet then makes new connection(s) and starts re-listing its ListWatches.
5. This puts a lot of load on the API server, which causes it to respond slowly, and takes us back to step 2.

In many tests the cycle of connection loss and re-creation would last for about 12 minutes. But sometimes it went on for a few hours.

The closing of connections in onRepeatedHeartbeatFailure was introduced to solve #48638, and it remains necessary for that purpose.  But, I don't think it needs to happen _repeatedly_ in a short space of time.  I'm about to submit a PR to address that. 

BTW, in case you're wondering why this doesn't happen in load tests with kubemark/hollow node, the reason is that `onRepeatedHeartbeatFailure` is `nil` in kubemark.  I spent a fruitless week testing with kubemark before tracking down the bug with 5000 real nodes :-)

--

I have verified all of the above, and the fact that is not due to failed pings, by building my own instrumented versions of kubelet and deploying them onto 5000 nodes.  There are no http ping failures. All the connection losses are driven by the sequence of events describe above.

**Comment by aojea** [comment]
I wonder if the kubelet should check the `/readyz` endpoint before start hammering the apiserver @wojtek-t 

Maybe this delays are caused because the endpoint is up but the apiserver is not ready?

**Comment by JohnRusk** [comment]
What happens if 5000 kubelets check readyz and then decide, all at about the same time, that the api server is ready, and start hammering it.  Aren't we back in the same situation then? ;-)

**Comment by JohnRusk** [comment]
As for the _cause_ of delays, they are just because the API server is busy.  Now... in theory, we could solve that by just giving it more CPU.  Eventually, if you give it enough cores to run on, that might reduce the timeouts sufficiently.  The problem with that approach, is that the necessary number of cores is at least a factor of 5 above what the cluster would otherwise need.  It's very inefficient to give the cluster >= 5x the CPU it "really" needs, just to cover the first few seconds after a restart.

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -595,7 +596,7 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
 		if closeAllConns == nil {
 			return errors.New("closeAllConns must be a valid function other than nil")
 		}
-		kubeDeps.OnHeartbeatFailure = closeAllConns
+		kubeDeps.OnHeartbeatFailure = createHeartbeatFailureHandler(closeAllConns, 5 * time.Minute)
```
Comment (liggitt) [inline cmd/kubelet/app/server.go:599]: is this still needed after https://github.com/kubernetes/kubernetes/pull/104844? that PR (in 1.23) switched to default to only closing idle connections, trusting the http/2 ping to close stuck open connections

**Comment by liggitt** [comment]
> * If two of those heartbeat operations fail in a row, kubelet calls a function named `onRepeatedHeartbeatFailure` which _closes all TCP connections_.

in 1.23, https://github.com/kubernetes/kubernetes/pull/104844 changed to only close idle connections by default, not all connections

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (JohnRusk) [reply cmd/kubelet/app/server.go:599]: Good question.  Checking my notes I see that I did one test with that change in place, on a single node out of the 5000, observed that the problem still happened on that node, and concluded that PR didn't help.  But, now that I understand the cause better, I see that I should have applied it to all nodes to improve the overall load pattern. (As I did later with the fix in this PR).  I will have to re-test with 104844.  Thanks @liggitt

**Comment by aojea** [comment]
> > * If two of those heartbeat operations fail in a row, kubelet calls a function named `onRepeatedHeartbeatFailure` which _closes all TCP connections_.
> 
> in 1.23, #104844 changed to only close idle connections by default, not all connections

Is kubelet.configured with the "is not rotating certificates" things? That caused to creat multiple connection to the apiservers per kubelet, ...

**Comment by JohnRusk** [comment]
Oh, good question @aojea. What exactly should I look for in the kubelet config?

**Comment by JohnRusk** [comment]
@aojea, I'll take a look around and figure out what to look for, in the kubelet settings, to check that. Will comment here once I'vye checked.

**Comment by aojea** [comment]
> Oh, good question @aojea. What exactly should I look for in the kubelet config?

> start kubelet with rotate-certificates=false


the scenario of the issue reported is very similar to the one you are describing https://github.com/kubernetes/kubernetes/issues/100849

> After PR is integrated, the connections between kubelet and kube-apiserver in our cluster with 4000 nodes increases by five times.

they went from 4k connections to 20k connections per kubelet , let's rule that out

**Comment by JohnRusk** [comment]
@aojea Just to be clear, do you mean "if rotatecerts=false then  #104844 will fix it" or do you mean "if rotatecerts=false then  #104844 will NOT fix it"?

**Comment by aojea** [comment]
> @aojea Just to be clear, do you mean "if rotatecerts=false then #104844 will fix it" or do you mean "if rotatecerts=false then #104844 will NOT fix it"?

if rotatecerts=false then #104844 will fix it

**Comment by JohnRusk** [comment]
Ah, OK. So I've been testing with rotatecerts=true.  So the problem is not fixed by the #104844 for me or for anyone else with rotatecerts=true. @liggitt , does that answer your question about whether this PR is needed?

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (JohnRusk) [reply cmd/kubelet/app/server.go:599]: See main comment thread for retest result and discusion. Short answer is yes, it is still needed :-)

**Comment by JohnRusk** [comment]
For completeness, I did a test with #104844 and rotatecerts = false.  As expected, that did fix the issue.

However, my understanding is that running with rotatecerts=false is not an attractive option for many users.  For instance, I'm thinking in particular of users using TLS bootstrapping. IIRC the bootstrapped certs expire after 12 months. So if you expect your cluster to last longer than a year, then I presume you want cert rotation otherwise won't the nodes stop working after 12 months?

Anyway, no matter what their reasons, many users will want to rotate certificates. So I think we need a solution that works for them.  Thinking about the discussion in this thread, the viable options seem to include either

(a) this PR, OR
(b) possibly building on what @liggitt said about relying on Pings for maintaining connection quality.  Would it be possible to rely solely on pings for detecting bad connections even when rotatecerts=true? I.e. reverse the work that was done for #48638 , leaving "close all connections" to be used _only_ for its original purpose of cert rotation and _not_ for handling bad connections?  This approach is perhaps more in keeping with @liggitt 's comments above, but it's also (IMHO) a bigger and therefore riskier change than this current PR.

**Comment by aojea** [comment]
> Anyway, no matter what their reasons, many users will want to rotate certificates. **So I think we need a solution that works when rotatecerts=true**. Thinking about the discussion in this thread, the viable options seem to include either

how can I reproduce it? you mention kubemark in some places ...

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (wojtek-t) [reply cmd/kubelet/app/server.go:599]: I guess I don't fully understand why that PR doesn't fix the problem (i.e. what exactly --rotate-certificates=true changes wrt to that). Can someone clarify that further? @aojea - FYI

[But assuming that indeed it's not addressed, this change makes sense to me.]

**Comment by JohnRusk** [comment]
@aojea It currently _cannot_ be reproduced with kubemark.  Unfortunately. (More on that below).  So I'm reproducing it an AKS cluster with 5000 nodes.  (I can get 5000 nodes, because I work in the AKS team ;-)

With 5000 nodes, the steps to repro are pretty simple: 
1. Set up the cluster
2. Optionally run something on it to create a bit of load. This is not essential. But often the effect seems to be more pronounced if the cluster is actually doing something (e.g. scheduling a bunch of pods) at step 3.  But I'd suggest you skip this step at first, for simplicity.
3. Restart the API server(s).
4. Watch API server CPU usage and request metrics (or kube-audit data).  Symptoms of problem are high CPU usage and lots of list requests with 429 responses.

Note that at step 4, if you see CPU spike to 5 or 10 times normal, but you _don't_ see the 429 responses, that probably means your CPU on the servers was over-provisioned _so much_ that it processed the load spike really quickly, and therefore didn't start causing many timeouts.  (I've never seen that happen, but I imagine it could happen in theory). In that case, you may need to reduce the sizing of your servers, so that the level of CPU over-provisioning is reduced.  E.g. try to force the issue to happen by provisioning only 4x the level of CPU that your cluster uses in its steady state (i.e. 4x it's level of use before step 3 above).  Or just simulate a reduction in CPU  by setting the GOMAXPROCS environment variable in your api servers to allow them to use fewer cores.  

What if you don't have 5000 real nodes ;-)   Then you'll need to fix kubemark first. Right now, kubemark has a "bug" in which it does not set the value of the `onRepeatedHeartbeatFailure` variable.  That's because kubelet init in kubemark goes through a somewhat different code path.  I had a look at fixing that myself but didn't have time to find a clean way to do it. In your case, if you just want a quick test, maybe you don't need a clean way!  Maybe you can just hack something together, to set `onRepeatedHeartbeatFailure` to the "close all" function in kubemark.  (The same close all function that real kubelet uses).  If you do that, I would expect you to be able to repro it with kubemark.  Just start up 5000 hollow nodes, then restart the API server.

**Comment by aojea** [comment]
so, when using the certificate rotation the transport is created here

https://github.com/kubernetes/kubernetes/blob/f49a15ae50196fc4d46d70f86de4d15cbc2405c6/cmd/kubelet/app/server.go#L848-L851

In this case, it seems the customer dialer is really needed to forcefully close the connections and cause the reconnection so the new certificate is used.

The same function used by the cert manager to close the connections is returned to use as callback and close the connections on the heartbeat client failure.

Some thoughts:
- Since HTTP2 this heartbeat hammer approach forcefully closing connections is not needed since each client-go will detect the stale connections and reconnection
- I don't know if there is a better way to force the TLS handshake though.
- I'd like to know how many connections are created by kubelet with rotate-certificates, since uses a custom dialer, but recently we had consolidated the transport for client-go in https://github.com/kubernetes/kubernetes/pull/105490

An immediate fix it is to replace the CloseConnection function returned and used for the heartbeat callback, by the CloseIdleConnections. 
However, I think that this code can be simplified now that we have http2 and the client-go transport is shared between types.

@wojtek-t @liggitt do we want to try to improve this or just going for the quick fix?

**Comment by liggitt** [comment]
> * I don't know if there is a better way to force the TLS handshake though.

Connections have to close and be re-established for a rotated client certificate to take effect

**Comment by aojea** [comment]
> > * I don't know if there is a better way to force the TLS handshake though.
> 
> Connections have to close and be re-established for a rotated client certificate to take effect

let me rephrase, a more "gracefully" to cause the reconnect than killing the TCP connection directly

**Comment by JohnRusk** [comment]
This PR is superceded by #108107
